File: core/IslandPerimeter/IslandPerimeter.py
from typing import List
from collections import deque


def islandPerimeter(grid: List[List[int]]) -> int:
    rows, cols = len(grid), len(grid[0])
    res = 0
    for row in range(rows):
        for col in range(cols):
            if grid[row][col] == 1:
                queue = deque([(row, col)])
                grid[row][col] = -1
                while queue:
                    step_row, step_col = queue.popleft()
                    for row_dir, col_dir in [[1, 0], [-1, 0], [0, 1], [0, -1]]:
                        new_row, new_col = step_row + row_dir, step_col + col_dir
                        if 0 <= new_row < rows and 0 <= new_col < cols:
                            if grid[new_row][new_col] == 1:
                                queue.append((new_row, new_col))
                                grid[new_row][new_col] = -1
                            elif grid[new_row][new_col] == 0:
                                res += 1
                        else:
                            res += 1
                return res

    return 0


if __name__ == '__main__':
    print(islandPerimeter([[1,1],[1,1]]))

File: core/IslandPerimeter/IslandPerimeter.java
package com.dbc;

public class IslandPerimeter {
    public int islandPerimeter(int[][] grid) {
        int res = 0;
        int[][] dirs = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
        for (int row = 0; row < grid.length; row++) {
            for (int col = 0; col < grid[0].length; col++) {
                if (grid[row][col] == 1) {
                    for (int[] dir : dirs) {
                        int new_row = dir[0] + row;
                        int new_col = dir[1] + col;
                        if (!(new_row >= 0 && new_row < grid.length && new_col >= 0 && new_col < grid[0].length && grid[new_row][new_col] == 1)) {
                            res++;
                        }
                    }
                }
            }
        }
        return res;
    }
}
